### DISCUSSION TITLE
Compilation of interesting szurubooru forks

### DISCUSSION THREAD

**Discussion by parallax4** [rr-/szurubooru#564] [General]
This project has a few different forks which contain a lot of changes and additions. I've been working on my own (not yet pushed) and have been borrowing bits and pieces from different forks, so I figured I'd share what I've found. If you're aware of any other forks with significant changes, please share them. I have a feeling there may be more lurking in the [Network](https://github.com/rr-/szurubooru/network) and [Forks](https://github.com/rr-/szurubooru/forks) pages.

* [Hunternif's fork](https://github.com/Hunternif/szurubooru) (@Hunternif) - Lots of minor and major changes, some of which were later integrated upstream but most of which weren't

* [infinibooru](https://github.com/VivaLaPanda/infinibooru) (@VivaLaPanda) - Stable Diffusion integration. Can generate permutations of existing images, I think

* [hujle's fork](https://github.com/hujle/szurubooru) (@hujle) - A bunch of changes that I'm still sifting through

* [szurubooru_photos](https://github.com/skybldev/szurubooru_photos) (@skybldev) - A fork aimed at storing photos and videos taken with physical cameras

**Comment by Hunternif** [comment]
Hey there, thanks for noting my fork.

I'll be happy to explain my changes in more detail if needed. There is definitely some technical debt there too, e.g. "similar posts" doesn't actually respect current safety settings.

I also showcase some of my changes in other posts here, e.g. https://github.com/rr-/szurubooru/discussions/434

**Reply by Kalidibus** [reply]
Do you know if your fork is directly compatible with the rr- version database files? I spent a while getting everything all tagged and imported and would like to not repeat that work if I need to switch in the future.

**Reply by Hunternif** [reply]
TLDR: it's not directly compatible.

My fork contains a few alembic revisions that create new tables. For my own purposes, I inserted my revision between some existing revisions. I think this means that when you merge my fork as-is, the old alembic revisions will not apply to your DB by default, and so my custom functionality will probably fail. I don't think it would affect your DB files, but best to check on a test setup.

However, some of my features are really esoteric and likely not very useful (e.g. metrics). I would advise to not merge those features, and instead cherrypick other more useful features, e.g. "similar posts" which don't modify the DB schema. But that would probably cause a lot of merge conflicts, sorry!

**Reply by Kalidibus** [reply]
All good appreciate the response!

**Reply by sticknuder** [reply]
Hi, @Hunternif, by chance can you explain how I could cherrypick your similar posts feature into my local copy of szurubooru? My copy is up to date with this repo plus some minor visual changes to some of the tpl files of the client code.

**Reply by sticknuder** [reply]
Disregard my above question. I managed to figure it out :D

**Comment by skybldev** [comment]
Hi! Thanks for including my fork.

FYI though, the fork is pretty much dead. I've implemented the features I wanted and that was pretty much that. The purpose of publishing it was more or less to get my changes out there and possibly merge it upstream, but it seems like it will be a while before that happens, and that's okay.

Time for my shameless future self-plug: I have ideas and plans (some already in motion) for a future booru that takes a lot of inspiration from this project. Just putting that out there in case anyone is interested.

**Reply by parallax4** [reply]
I considered making my own as well but it's such a huge undertaking to create a big project from scratch that I've just decided to heavily mod szurubooru. I'll share my fork within a month or two, most likely.

Have you made significant progress on your project? It can be easy to underestimate just how much work it takes to start something new like this.

**Reply by skybldev** [reply]
Haha, underestimation is its middle name! Although, I'm not planning to make this some production-level thing right away. It's more of a passion project that I can apply thing I have learned from college as I go.

So far the most significant thing that has happened is that I have laid out the groundwork in the form of the database structure. I have also planned out the UI, emphasizing simplicity and orthogonality. The only thing left is to design the API andpre-document before I actually start building it.

My justifications for starting from scratch are: I don't want to build upon old code, I absolutely rue every moment I have to use Python, and there are already many great examples of boorus to base mine off of, and then add some more.

**Reply by parallax4** [reply]
True, szurubooru really is ancient code. What language are you writing yours in?

**Reply by skybldev** [reply]
I'm writing mine in Rust for the backend and TS+Deno

**Comment by hujle** [comment]
Appreciate the mentioning, but my fork is just an attempt to combine OG szurubooru with another fork of it that contains OpenGraph embeds, there's nothing interesting about it.

**Reply by parallax4** [reply]
Can you describe the objectives/advantages of this? (I don't know much about OpenGraph.)

**Reply by hujle** [reply]
I don't think i can, but it basically allows websites and apps to generate embeds (like this in case of Discord). Quite sad that original repo still doesn't have proper embeds implementation. I should probably create an issue about it, bringing it up once again.
![изображение](https://user-images.githubusercontent.com/16465620/231079977-c9f3c505-77a7-42ee-af83-c018bd199945.png)

**Reply by hujle** [reply]
Here are #441 and #422 if you're interested, but both are quite outdated and doesn't work properly with current iteration of repo (or is it just me doing it wrong).

**Reply by sgsunder** [reply]
I'll try and revisit #441 when I get a chance. Still feel like it's a bit hacky at the moment.

**Comment by po5** [comment]
I have been fixing bugs and layout issues I come across while setting up my instance, and implementing features I am missing.  
https://github.com/po5/szurubooru/tree/vb

Notable changes are instant theme application on reload, correct resizing on desktop and android, gallery-dl support, ruffle for swf playback, higher quality and lower filesize thumbnails (with better swf support), right click support on firefox (that keeps notes functionality), markdown header support, original aspect ratio thumbnails, style fixes, stripes instead of transparency grid, server-agnostic download button, thumbnails as image preloads.

Some commits are personal preference/removing things I don't need, like 3rd party reverse search links.  
I have merged some features from other forks, the source is labeled in extended commit messages.

I plan to upstream all of the interesting changes, but that will only happen once I'm done messing with it every day.  
Feature todo list here, random fixes not listed https://github.com/users/po5/projects/1

**Reply by pelicans45** [reply]
@po5 Hi, I noticed some changes you made that seem to suggest you replaced waitress with fastwsgi. Do you find it to be more performant? Do you foresee any issues moving away from waitress, with its worker threads? (There's a lot of discussion/documentation in szurubooru about waitress's worker threads being helpful for background tasks.)

**Reply by po5** [reply]
Some issues like everything locking up during an upload did not seem to be fixed when increasing waitress worker count. I also ran into annoying desyncs a few times where data would be different from request to request, can't recall if that was waitress's fault or my own. I switched to fastwsgi as a shot in the dark and while I don't think it's doing any better, it doesn't seem to cause any issues. Site is running smoothly :^) Please share any findings if you mess with it more than I did.

**Comment by liamw1** [comment]
A few months ago, I started a complete rewrite of the szurubooru backend using Rust with a focus on increasing performance: [https://github.com/liamw1/oxibooru](https://github.com/liamw1/oxibooru)

I've implemented maybe 90% of the original functionality so far and have added a few notable improvements, including:
- Faster and more thorough reverse-image search
- Improved post signature indexing
- Better cyclic tag dependency detection
- 'Getting around' posts made to respect sorting
- More sort-style tokens for posts
- Better load times for post search by reducing number of SQL round-trips
- Probably some other things I forgot

There's still a few bugs to work out, but it works pretty well right now for my own personal use. I managed to manually migrate an old szurubooru database over, but I plan to create a script for this when I get around to it.

EDIT: I put together some rough benchmarks against szurubooru [here](https://github.com/liamw1/oxibooru/blob/master/doc/BENCHMARKS.md).

**Reply by skybldev** [reply]
oh my god!! the dream come true! very impressive work 🥳

**Reply by gljames24** [reply]
This is exactly what I've wanted since I started migrating away from hydrus to szurubooru. This and moving my booru to ssds is going to really speed it up. I also know Rust a lot better than Python so I might be able to contribute if I have time.

**Reply by parallax4** [reply]
@liamw1 Awesome. Do you have a demo instance you could link? I've also made my own fork (currently not public) with some performance improvements, and I'd be interested in comparing how the two feel. One big change I added is making the tag autocomplete a lot faster.

**Reply by liamw1** [reply]
@parallax4 I don't have a demo instance at the moment. At some point I'll finish the conversion script to convert a szurubooru database to mine and you could compare a converted database. It definitely _feels_ much faster. I'll run some benchmarks when I get the chance so you can have a more quantitative comparison.

**Reply by parallax4** [reply]
@liamw1 That's great. Odds are mine isn't nearly as fast as yours; it's just a bit faster than base szurubooru. If you're interested you can see the booru at uggcf://wrw.yby (rot13'd).

**Reply by liamw1** [reply]
@parallax4 I actually just pushed some benchmark results I threw together: [link](https://github.com/liamw1/oxibooru/blob/master/doc/BENCHMARKS.md). If there's something that I didn't include that you'd think would be interesting to benchmark let me know and I can add it.

**Reply by po5** [reply]
> Looking at the client logs, the client actually makes two info requests when viewing the homepage.

This is one of the many bugs fixed in my fork, if you want to port over the client improvements. https://github.com/po5/szurubooru/commit/59ad82512b05829c1b8c870f22c00cc0abffd1e4

**Reply by parallax4** [reply]
@po5 I noticed that too and I just got rid of the info endpoint entirely in mine and embedded it in the JavaScript. I just rebuild the container if I update the config. Makes the initial loading a bit faster.

**Comment by pelicans45** [comment]
My fork: https://github.com/pelicans45/glebooru

Link to booru instance: https://jej.lol

Has a bunch of tweaks and design changes

**Reply by kemopup** [reply]
ouch! that edge thou

**Reply by pelicans45** [reply]
What do you mean?
